vowpalwabbit/shared_data.cc

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#include <iomanip>
+
+#include "shared_data.h"
+#include "memory.h"
+
+#include <cfloat>
+#include <climits>
+
+shared_data::shared_data()
+    : is_more_than_two_labels_observed(false), first_observed_label(FLT_MAX), second_observed_label(FLT_MAX)
+{
+}
+
+shared_data::shared_data(const shared_data& other)
+{
+  queries = other.queries;
+  example_number = other.example_number;
+  total_features = other.total_features;
+  t = other.t;
+  weighted_labeled_examples = other.weighted_labeled_examples;
+  old_weighted_labeled_examples = other.old_weighted_labeled_examples;
+  weighted_unlabeled_examples = other.weighted_unlabeled_examples;
+  weighted_labels = other.weighted_labels;
+  sum_loss = other.sum_loss;
+  sum_loss_since_last_dump = other.sum_loss_since_last_dump;
+  dump_interval = other.dump_interval;
+  gravity = other.gravity;
+  contraction = other.contraction;
+  min_label = other.min_label;
+  max_label = other.max_label;
+  if (other.ldict) { ldict = VW::make_unique<VW::named_labels>(*other.ldict); }

```

Comment:
Unfortunately we have to define the copy constructor due to the unique ptr and the atomics. Luckily we get move for free!

Reply:
can't we do: `ldict = std::move(other.ldict)` here?

Reply:
nevermind -> copy constructor

Reply:
Ah it seems like this is not okay on MacOS :(